README.md
# Rust OpenCV bindings

[![Build status](https://travis-ci.org/twistedfall/opencv-rust.svg?branch=master)](https://travis-ci.org/twistedfall/opencv-rust)
[![Documentation](https://docs.rs/opencv/badge.svg)](https://docs.rs/opencv)
[![Package](https://img.shields.io/crates/v/opencv.svg)](https://crates.io/crates/opencv)

Experimental Rust bindings for OpenCV 3 and 4.

The API is usable but unstable and not very battle-tested; use at your own risk.

todo: rearrange, bugreporting guidelines (-vv, rust_backtrace)

## API

[API Documentation](https://docs.rs/opencv) is, to varying success,
translated from OpenCV's doxygen docs. Most likely you'll want to
refer to the official [OpenCV C++
documentation](https://docs.opencv.org/master/) as well.

### OpenCV version support

The following OpenCV versions are supported at the moment:
* 3.2
* 3.4
* 4.2

You can choose the target OpenCV version with the following Cargo features:
* `opencv-32`
* `opencv-34`
* `opencv-4` (default)

If you need support for `contrib` modules, also enable `contrib` feature.

### Quickstart

Make sure the supported OpenCV version is installed in your system along with `pkg_config`
files or `-dev` packages.

Update your Cargo.toml
```toml
opencv = "0.29"
```

Select OpenCV version if different from default in Cargo.toml:
```toml
opencv = {version = "0.29", default-features = false, features = ["opencv-34"]}
```

And enable usage of `contrib` modules:
```toml
opencv = {version = "0.29", features = ["contrib"]}
```

Import prelude
```rust
use opencv::prelude::*;
```

### Minimum rustc version

Generally you should use the latest stable rustc to compile this crate.

### Platform support

Currently the main development and testing of the crate is performed on Linux, but other 
major platforms are also supported: Mac OS X and Windows.

Support for Windows is a bit tricky at the moment: you will need to set up `OPENCV_LINK_LIBS`,
`OPENCV_LINK_PATHS` and `OPENCV_INCLUDE_PATHS` environment
variables. Then build the crate with `buildtime-bindgen` feature enabled. If you get linking
errors try also building with `--release` flag.

Mac OS X build currently also requires `buildtime-bindgen` feature enabled.

Also refer to the corresponding [issue #6](https://github.com/twistedfall/opencv-rust/issues/6), 
[issue #70](https://github.com/twistedfall/opencv-rust/issues/70) and Travis
[build script](https://github.com/twistedfall/opencv-rust/blob/master/ci/script.sh).

### Features
* `opencv-32` - build against OpenCV 3.2.0, this feature is aimed primarily on stable Debian and
  Ubuntu users who can install OpenCV from the repository without having to compile it from the
  source
* `opencv-34` - build against OpenCV 3.4.x
* `opencv-4` (default) - build against OpenCV 4.x
* `contrib` - enable the usage of OpenCV contrib modules for corresponding OpenCV version
* `buildtime-bindgen` - regenerate all bindings, should only be used during the crate development
  or when building on Windows or Mac OS X, with this feature enabled the bundled headers are no
  longer used for the code generation, the ones from the installed OpenCV are
* `docs-only` - internal usage, for building docs on [docs.rs](https://docs.rs/opencv)

### Functionality

Generally the crate tries to only wrap OpenCV API and provide some convenience functions
to be able to use it in Rust easier. We try to avoid adding any functionality besides
that.

### Callbacks

Some API functions accept callbacks, e.g. set_mouse_callback. While currently
it's possible to successfully use those functions there are some limitations to
keep in mind. Current implementation of callback handling keeps hold of the
passed callback argument forever. That means that the closure used as a callback
will never be freed during the lifetime of a program and moreover Drop will
not be called for it (they are stored in global static [`Slab`](https://crates.io/crates/slab)).
There is a plan to implement possibility to be able to free at least some of the
closures.

### Unsafety

Although crate tries to provide ergonomic Rust interface for OpenCV, don't expect
Rust safety guarantees at this stage. It's especially true for borrow checking and
shared mutable ownership. Notable example would be `Mat` which is a reference counted
object in its essence. You can own a seemingly separate `Mat` in Rust terms, but
it's going to be a mutable reference to the other `Mat` under the hood. Treat safety
of the crate's API as you would treat one of C++, use `clone()` when needed.

## Getting Started

The following external dependencies are required:
- python3
- OpenCV 3.2.0, 3.4.9 or 4.2.0

OpenCV is a complicated dependency with a lot of different
configurations that can break this crate since it relies on fragile
header parsing. Try these strategies in order, continuing to the next
if builds with this crate fail.

### See if you already have the right version of OpenCV installed

if OpenCV is installed, its version can be checked with:

```sh
opencv_version
```

### Install OpenCV

Install supported version of OpenCV. Check your platform's package manager or see the
upstream OpenCV
[installation guides](https://docs.opencv.org/master/df/d65/tutorial_table_of_content_introduction.html).

When your OpenCV is installed in a location that's not being picked up by build script
you can try setting one or both of the following environment vars prior to trying to
build the crate:

```bash
export PKG_CONFIG_PATH=/custom/prefix/lib/pkgconfig/
export LD_LIBRARY_PATH=/custom/prefix/lib/
```

Path specified by `PKG_CONFIG_PATH` must contain `opencv.pc` or `opencv4.pc` (for OpenCV 4.x).
Path specified by `LD_LIBRARY_PATH` must contain `libopencv_*.so` files.

### Environment variables

* `OPENCV_HEADER_DIR`
  During crate build it uses OpenCV headers supplied inside the crate for binding generation.
  If you want to use your own (system) headers supply `OPENCV_HEADER_DIR` environment variable.
  The directory in that environment variable should contain `opencv2` dir, e.g. `/usr/include` for
  OpenCV-3.4.x or `/usr/include/opencv4` for OpenCV-4.x.

* `OPENCV_PACKAGE_NAME`
  In some cases you might want to override the pkgconfig or vcpkg package name, you can use this environment
  variable for that. If you set it pkgconfig will expect to find the file with that name and `.pc` extension
  in the package directory. And vcpkg will use that name to try to find package in `packages` directory under
  `VCPKG_ROOT`. For compatibility reasons `OPENCV_PKGCONFIG_NAME` is also supported as variable name.

The following variables must be set when building without `pkg_config` or `vcpkg`

CHANGES.md
* 0.31
  * libclang
  * core::sum() → core::sum_elems()
  * cv_return_value → Result
  * _ref → ref_
  * VectorOfuchar → VectorOfu8

* 0.30
  * data_typed() and data_typed_mut() now do is_continuous() check

* 0.29.2
  * add ability to get reference to the inner value of non-trait `PtrOf*` types (kudos to jerry73204)

* 0.29.1
  * fix generation of setters for properties (kudos to jerry73204)

* 0.29
  * default features now target OpenCV 4.2.0 (instead of OpenCV 3.4.x). Feature `opencv-42` was renamed to
    `opencv-4`. Bindings were bumped to OpenCV 4.2.0 and 3.4.9.

  * core::MAKETYPE was renamed to CV_MAKETYPE, core::MAT_DEPTH was renamed to CV_MAT_DEPTH

* 0.28
  * dnn::BackendNode constructor as well as stitching::createStitcher and stitching::createStitcherScans are no
    longer generated for the OpenCV version where they are not available, fixing the linking issues

* 0.27.0
  * add support for `vcpkg` when building for windows, it's being used by default in the absence of
    environment variables
    
  * with `buildtime-bidngen` feature enabled the crate no longer uses bundled source files for code
    generation by default, but the ones installed with OpenCV library detected by `pkg_config`, `vcpkg` or
    environment; set `OPENCV_HEADER_DIR` environment variable to override this behavior  

* 0.26.6
  * ...


build.rs
// todo check if build with system headers is successful

use std::{
	borrow::Cow,
	collections::HashSet,
	env,
	fs::{self, File},
	io::{BufRead, BufReader},
	iter::FromIterator,
	path::{Path, PathBuf},
};

use glob_crate::glob;
use once_cell::sync::{Lazy, OnceCell};
use semver::{Version, VersionReq};

#[cfg(feature = "buildtime-bindgen")]
mod generator {
	use std::{
		collections::HashSet,
		ffi::OsStr,
		fs::{self, DirEntry, File, OpenOptions},
		io::{self, BufRead, BufReader, Write},
		iter::FromIterator,
		path::{Path, PathBuf},
		time::Instant,
	};

	use glob_crate::glob;
	use once_cell::sync::Lazy;
	use rayon::prelude::*;

	use super::{
		file_copy_to_dir,
		get_versioned_hub_dirs,
		MODULES,
		OUT_DIR,
		Result,
		SRC_CPP_DIR,
		SRC_DIR,
	};

	static CORE_MODULES: Lazy<HashSet<&'static str>> = Lazy::new(|| HashSet::from_iter([
		"calib3d",
		"core",
		#[cfg(not(feature = "opencv-32"))]
		"dnn",
		#[cfg(feature = "opencv-4")]
		"dnn_superres",
		"features2d",
		"flann",
		#[cfg(feature = "opencv-4")]
		"gapi",
		"highgui",
		"imgcodecs",
		"imgproc",
		"ml",
		"objdetect",
		"photo",
		#[cfg(any(feature = "opencv-32", feature = "opencv-34"))]
		"shape",
		"stitching",
		#[cfg(any(feature = "opencv-32", feature = "opencv-34"))]
		"superres",
		"video",
		"videoio",
		#[cfg(any(feature = "opencv-32", feature = "opencv-34"))]
		"videostab",
		"viz",
	].iter().copied()));

	fn read_dir(path: &Path) -> Result<impl Iterator<Item=DirEntry>> {
		Ok(path.read_dir()?.filter_map(|e| e.ok()))
	}

	fn copy_indent(mut read: impl BufRead, mut write: impl Write, indent: &str) -> Result<()> {
		let mut line = Vec::with_capacity(100);
		while read.read_until(b'\n', &mut line)? != 0 {
			write.write(indent.as_bytes())?;
			write.write(&line)?;
			line.clear();
		}
		Ok(())
	}

	fn file_move_to_dir(src_file: &Path, target_dir: &Path) -> Result<PathBuf> {
		if !target_dir.exists() {
			fs::create_dir_all(&target_dir)?;
		}
		let src_filename = src_file.file_name()
			.ok_or_else(|| "Can't calculate filename")?;
		let target_file = target_dir.join(src_filename);
		if fs::rename(&src_file, &target_file).is_err() {
			fs::copy(&src_file, &target_file)?;
			fs::remove_file(src_file)?;
		}
		Ok(target_file)
	}

	pub fn gen_wrapper(opencv_header_dir: &Path) -> Result<()> {
		println!("cargo:rerun-if-env-changed=OPENCV_HEADER_DIR");
		println!("cargo:rerun-if-env-changed=OPENCV_PACKAGE_NAME");
		println!("cargo:rerun-if-env-changed=OPENCV_PKGCONFIG_NAME");
		println!("cargo:rerun-if-env-changed=OPENCV_LINK_LIBS");
		println!("cargo:rerun-if-env-changed=OPENCV_LINK_PATHS");
		println!("cargo:rerun-if-env-changed=OPENCV_INCLUDE_PATHS");

		let include_exts = &[OsStr::new("rs"), OsStr::new("cpp"), OsStr::new("hpp"), OsStr::new("toml"), OsStr::new("lock")];
		for entry in SRC_CPP_DIR.read_dir()?.map(|e| e.unwrap()) {
			let path = entry.path();
			if path.is_file() && path.extension().map_or(false, |e| include_exts.contains(&e)) {
				if let Some(path) = path.to_str() {
					println!("cargo:rerun-if-changed={}", path);
				}
			}
		}
		let out_dir_as_str = OUT_DIR.to_str().unwrap();
		let (rust_hub_dir, cpp_hub_dir) = get_versioned_hub_dirs();
		let module_dir = rust_hub_dir.join("hub");
		let manual_dir = SRC_DIR.join("manual");
		let opencv_dir = opencv_header_dir.join("opencv2");

		eprintln!("=== Using OpenCV headers from: {}", opencv_dir.display());
		eprintln!("=== Generating code in: {}", out_dir_as_str);
		eprintln!("=== Placing generated bindings into: {}", rust_hub_dir.display());

		let modules = MODULES.get().expect("MODULES not initialized");

		for entry in read_dir(&OUT_DIR)? {
			let path = entry.path();
			if path.is_file() {
				let _ = fs::remove_file(path);
			}
		}

		let version = if cfg!(feature = "opencv-32") {
			"3.2.0"
		} else if cfg!(feature = "opencv-34") {
			"3.4.9"
		} else if cfg!(feature = "opencv-4") {
			"4.2.0"
		} else {
			unreachable!();
		};
		let clang = clang::Clang::new().expect("Cannot initialize clang");
		let start = Instant::now();
		modules.par_iter().for_each(|module| {
			let mut module_file = SRC_CPP_DIR.join(format!("{}.hpp", module));
			if !module_file.exists() {
				module_file = opencv_header_dir.join(format!("opencv2/{}.hpp", module));
			}
			let bindings_writer = binding_generator::writer::RustBindingWriter::new(
				opencv_header_dir,
				&*SRC_CPP_DIR,
				&*OUT_DIR,
				&module,
				version,
				false,
			);
			binding_generator::Generator::new(&opencv_header_dir, &*SRC_CPP_DIR, module, &clang)
				.process_file(&module_file, bindings_writer);
			println!("Generated: {}", module);
		});
		println!("Total binding generation time: {:?}", start.elapsed());

		if !module_dir.exists() {
			fs::create_dir_all(&module_dir)?;
		}

		for entry in read_dir(&module_dir)? {
			let path = entry.path();
			if path.extension().map_or(false, |e| e == "rs") {
				let _ = fs::remove_file(path);
			}
		}

		if !cpp_hub_dir.exists() {
			fs::create_dir_all(&cpp_hub_dir)?;
		}
		for entry in read_dir(&cpp_hub_dir)? {
			let path = entry.path();
			if path.is_file() {
				let _ = fs::remove_file(path);
			}
		}

		for module in modules {
			let module_cpp = OUT_DIR.join(format!("{}.cpp", module));
			if module_cpp.is_file() {
				file_copy_to_dir(&module_cpp, &cpp_hub_dir)?;
				let module_types_cpp = OUT_DIR.join(format!("{}_types.hpp", module));
				let mut module_types_file = OpenOptions::new().create(true).truncate(true).write(true).open(&module_types_cpp)?;
				let mut type_files: Vec<PathBuf> = glob(&format!("{}/???-{}-*.type.cpp", out_dir_as_str, module))?
					.collect::<Result<_, glob_crate::GlobError>>()?;
				type_files.sort_unstable();
				for entry in type_files.into_iter() {
					io::copy(&mut File::open(entry)?, &mut module_types_file)?;
				}
				file_copy_to_dir(&module_types_cpp, &cpp_hub_dir)?;
			}
		}

		let add_manual = |file: &mut File, mod_name: &str| -> Result<bool> {
			if manual_dir.join(format!("{}.rs", mod_name)).exists() {
				writeln!(file, "pub use crate::manual::{}::*;", mod_name)?;
				Ok(true)
			} else {
				Ok(false)
			}
		};

		{
			let mut hub_rs = File::create(rust_hub_dir.join("hub.rs"))?;

			let mut types_rs = File::create(module_dir.join("types.rs"))?;
			writeln!(&mut types_rs, "use crate::{{mod_prelude::*, core, types, sys}};")?;
			writeln!(&mut types_rs, "")?;

			let mut sys_rs = File::create(module_dir.join("sys.rs"))?;
			writeln!(&mut sys_rs, "use crate::{{mod_prelude_types::*, core}};")?;
			writeln!(&mut sys_rs, "")?;
			for module in modules {
				let is_contrib_module = !CORE_MODULES.contains(module.as_str());
				let write_if_contrib = |write: &mut File| -> Result<()> {
					if is_contrib_module {
						writeln!(write, r#"#[cfg(feature = "contrib")]"#)?;
					}
					Ok(())
				};
				// hub
				write_if_contrib(&mut hub_rs)?;
				writeln!(&mut hub_rs, "pub mod {};", module)?;
				let module_filename = format!("{}.rs", module);
				let target_file = file_move_to_dir(&OUT_DIR.join(&module_filename), &module_dir)?;
				let mut f = OpenOptions::new().append(true).open(&target_file)?;
				add_manual(&mut f, module)?;

				// types
				let mut write_header = true;
				for entry in glob(&format!("{}/???-{}-*.type.rs", out_dir_as_str, module))? {
					let entry = entry?;
					if entry.metadata().map(|meta| meta.len()).unwrap_or(0) > 0 {
						if write_header {
							write_if_contrib(&mut types_rs)?;
							writeln!(&mut types_rs, "mod {}_types {{", module)?;
							writeln!(&mut types_rs, "\tuse super::*;")?;
							writeln!(&mut types_rs, "")?;
							write_header = false;
						}
						copy_indent(BufReader::new(File::open(&entry)?), &mut types_rs, "\t")?;
					}
				}
				if !write_header {
					writeln!(&mut types_rs, "}}")?;
					write_if_contrib(&mut types_rs)?;
					writeln!(&mut types_rs, "pub use {}_types::*;", module)?;
					writeln!(&mut types_rs, "")?;
				}

				// sys
				let path = OUT_DIR.join(format!("{}.externs.rs", module));
				write_if_contrib(&mut sys_rs)?;
				writeln!(&mut sys_rs, "mod {}_sys {{", module)?;
				writeln!(&mut sys_rs, "\tuse super::*;")?;
				writeln!(&mut sys_rs, "")?;
				for entry in glob(&format!("{}/{}-*.rv.rs", out_dir_as_str, module))? {
					let entry: PathBuf = entry?;
					copy_indent(BufReader::new(File::open(entry)?), &mut sys_rs, "\t")?;
				}
				copy_indent(BufReader::new(File::open(&path)?), &mut sys_rs, "\t")?;
				writeln!(&mut sys_rs, "}}")?;
				write_if_contrib(&mut sys_rs)?;
				writeln!(&mut sys_rs, "pub use {}_sys::*;", module)?;
				writeln!(&mut sys_rs, "")?;
			}
			writeln!(&mut hub_rs, "pub mod types;")?;
			writeln!(&mut hub_rs, "#[doc(hidden)]")?;
			writeln!(&mut hub_rs, "pub mod sys;")?;

			add_manual(&mut types_rs, "types")?;

			add_manual(&mut sys_rs, "sys")?;

			writeln!(&mut hub_rs, "pub mod hub_prelude {{")?;
			for module in modules {
				if !CORE_MODULES.contains(module.as_str()) {
					writeln!(&mut hub_rs, r#"	#[cfg(feature = "contrib")]"#)?;
				}
				writeln!(&mut hub_rs, r#"	pub use super::{}::prelude::*;"#, module)?;
			}
			writeln!(&mut hub_rs, "}}")?;
		}

		Ok(())
	}
}

type Result<T, E = Box<dyn std::error::Error>> = std::result::Result<T, E>;

static MODULES: OnceCell<Vec<String>> = OnceCell::new();

static OUT_DIR: Lazy<PathBuf> = Lazy::new(|| PathBuf::from(env::var_os("OUT_DIR").expect("Can't read OUT_DIR env var")));
static MANIFEST_DIR: Lazy<PathBuf> = Lazy::new(|| PathBuf::from(env::var_os("CARGO_MANIFEST_DIR").expect("Can't read CARGO_MANIFEST_DIR env var")));
static SRC_DIR: Lazy<PathBuf> = Lazy::new(|| MANIFEST_DIR.join("src"));
static SRC_CPP_DIR: Lazy<PathBuf> = Lazy::new(|| MANIFEST_DIR.join("src_cpp"));

#[derive(Debug)]
struct Library {
	pub pkg_name: String,
	pub libs: Vec<String>,
	pub link_paths: Vec<PathBuf>,
	pub framework_paths: Vec<PathBuf>,
	pub include_paths: Vec<PathBuf>,
	pub version: String,
	pub prefix: PathBuf,
	pub libdir: PathBuf,
	pub cargo_metadata: Vec<String>,
}

impl Library {
	pub fn probe_from_paths(pkg_name: &str, link_libs: &str, link_paths: &str, include_paths: &str) -> Result<Self> {
		eprintln!(
			"=== Setting up OpenCV library from environment, pkg_name: {}, link_libs: {}, link_paths: {}, include_paths: {}",
			pkg_name, link_libs, link_paths, include_paths
		);
		let mut cargo_metadata = Vec::with_capacity(64);
		let libs: Vec<_> = link_libs.split(',')
			.map(|x| x.trim())
			.filter(|x| !x.is_empty())
			.map(|x| {
				let mut path = PathBuf::from(x);
				path.set_extension("");
				let mut out = path.file_name()
					.and_then(|f| f.to_str()).expect("Invalid library name")
					.to_owned();
				cargo_metadata.push(format!("cargo:rustc-link-lib={}", out));
				if cfg!(target_env = "msvc") {
					out += ".lib";
				}
				out
			})
			.collect();

		let link_paths: Vec<_> = link_paths.split(',')
			.map(|x| x.trim())
			.filter(|x| !x.is_empty())
			.map(|x| {
				let out = PathBuf::from(x);
				cargo_metadata.push(
					format!("cargo:rustc-link-search=native={}", out.to_str().expect("Invalid link path"))
				);
				out
			})
			.collect();

		let libdir = link_paths
			.first()
			.map_or_else(|| PathBuf::from(""), |x| x.clone());

		let include_paths: Vec<_> = include_paths.split(',')
			.map(|x| x.trim())
			.filter(|x| !x.is_empty())
			.map(PathBuf::from)
			.collect();

		let version = include_paths.iter()
			.filter_map(get_version_from_headers)
			.next();

		let out = Self {
			pkg_name: pkg_name.to_owned(),
			libs,
			link_paths,
			framework_paths: vec![],
			include_paths,
			version: version.unwrap_or("0.0.0".to_owned()),
			prefix: PathBuf::from(""),
			libdir,
			cargo_metadata,
		};
		Ok(out)
	}

	#[cfg(not(target_env = "msvc"))]
	pub fn probe_system(pkg_name: &str) -> Result<Self> {
		eprintln!("=== Setting up OpenCV library from pkg_config");
		let mut config = pkg_config::Config::new();
		config.cargo_metadata(false);
		let opencv = config.probe(pkg_name)?;
		let mut cargo_metadata = Vec::with_capacity(64);
		cargo_metadata.extend(
			opencv.link_paths.iter()
				.map(|link_path| format!("cargo:rustc-link-search=native={}", link_path.to_str().expect("Invalid link path")))
				.chain(opencv.libs.iter().map(|lib| format!("cargo:rustc-link-lib={}", lib))));
		Ok(Self {
			pkg_name: pkg_name.to_owned(),
			libs: opencv.libs,
			link_paths: opencv.link_paths,
			framework_paths: opencv.framework_paths,
			include_paths: opencv.include_paths,
			version: opencv.version,
			prefix: PathBuf::from(pkg_config::get_variable(pkg_name, "prefix")?),
			libdir: PathBuf::from(pkg_config::get_variable(pkg_name, "libdir")?),
			cargo_metadata,
		})
	}

	#[cfg(target_env = "msvc")]
	pub fn probe_system(pkg_name: &str) -> Result<Self> {
		eprintln!("=== Setting up OpenCV library from vcpkg");
		let mut config = vcpkg::Config::new();
		config.cargo_metadata(false);
		let opencv = config.find_package(pkg_name)?;
		let libs = opencv.found_libs.into_iter()
			.filter_map(|lib| lib.file_name()
				.and_then(|l| l.to_str())
				.map(|l| l.to_string())
			)
			.collect();
		let version = opencv.include_paths.iter()
			.filter_map(get_version_from_headers)
			.next();
		let libdir = opencv.link_paths.first()
			.map_or_else(|| PathBuf::from(""), |x| x.clone());

		Ok(Self {
			pkg_name: pkg_name.to_owned(),
			libs,
			link_paths: opencv.link_paths,
			framework_paths: vec![],
			include_paths: opencv.include_paths,
			version: version.unwrap_or("0.0.0".to_owned()),
			prefix: PathBuf::from(""),
			libdir,
			cargo_metadata: opencv.cargo_metadata,
		})
	}

	pub fn probe() -> Result<Self> {
		let env_vars = (env::var("OPENCV_LINK_LIBS"), env::var("OPENCV_LINK_PATHS"), env::var("OPENCV_INCLUDE_PATHS"));
		let env_vars = if let (Ok(link_libs), Ok(link_paths), Ok(include_paths)) = env_vars {
			Some((link_libs, link_paths, include_paths))
		} else {
			None
		};
		let pkg_name = if cfg!(feature = "opencv-32") || cfg!(feature = "opencv-34") {
			env::var("OPENCV_PACKAGE_NAME")
				.or_else(|_| env::var("OPENCV_PKGCONFIG_NAME"))
				.map(|x| Cow::Owned(x))
				.unwrap_or_else(|_| if cfg!(target_env = "msvc") && env_vars.is_none() {
					"opencv3".into() // the package name in vcpkg for OpenCV3.x is different: https://github.com/microsoft/vcpkg/tree/master/ports/opencv3
				} else {
					"opencv".into()
				})
		} else if cfg!(feature = "opencv-4") {
			env::var("OPENCV_PACKAGE_NAME")
				.or_else(|_| env::var("OPENCV_PKGCONFIG_NAME"))
				.map(|x| Cow::Owned(x))
				.unwrap_or_else(|_| "opencv4".into())
		} else {
			unreachable!("Feature flags should have been checked in main()");
		};
		if let Some((link_libs, link_paths, include_paths)) = env_vars {
			Self::probe_from_paths(pkg_name.as_ref(), &link_libs, &link_paths, &include_paths)
		} else {
			Self::probe_system(pkg_name.as_ref())
		}.map_err(|e| format!("Package {} is not found, caused by: {}", pkg_name, e).into())
	}

	pub fn emit_cargo_metadata(&self) {
		self.cargo_metadata.iter().for_each(|meta| {
			println!("{}", meta);
		});
	}
}

fn file_copy_to_dir(src_file: &Path, target_dir: &Path) -> Result<PathBuf> {
	if !target_dir.exists() {
		fs::create_dir_all(&target_dir)?;
	}
	let src_filename = src_file.file_name()
		.ok_or_else(|| "Can't calculate filename")?;
	let target_file = target_dir.join(src_filename);
	fs::copy(&src_file, &target_file)?;
	Ok(target_file)
}

fn get_version_from_headers(header_dir: &PathBuf) -> Option<String> {
	let version_hpp = header_dir.join("opencv2/core/version.hpp");
	if !version_hpp.is_file() {
		return None;
	}
	let mut major = None;
	let mut minor = None;
	let mut revision = None;
	let mut line = String::with_capacity(256);
	let mut reader = BufReader::new(File::open(version_hpp).ok()?);
	while let Ok(bytes_read) = reader.read_line(&mut line) {
		if bytes_read == 0 {
			break;
		}
		const PREFIX: &str = "#define CV_VERSION_";
		if line.starts_with(PREFIX) {
			let mut parts = line[PREFIX.len()..].split_whitespace();
			if let (Some(ver_spec), Some(version)) = (parts.next(), parts.next()) {
				match ver_spec {
					"MAJOR" => {
						major = Some(version.to_string());
					}
					"MINOR" => {
						minor = Some(version.to_string());
					}
					"REVISION" => {
						revision = Some(version.to_string());
					}
					_ => {}
				}
			}
			if major.is_some() && minor.is_some() && revision.is_some() {
				break;
			}
		}
		line.clear();
	}
	if let (Some(major), Some(minor), Some(revision)) = (major, minor, revision) {
		Some(format!("{}.{}.{}", major, minor, revision))
	} else {
		Some("0.0.0".to_string())
	}
}

fn check_matching_version(version: &str) -> Result<()> {
	if cfg!(feature = "opencv-32") && !VersionReq::parse("~3.2")?.matches(&Version::parse(version)?) {
		Err(format!("OpenCV version: {} must be from 3.2 branch because of the feature: opencv-32", version).into())
	} else if cfg!(feature = "opencv-34") && !VersionReq::parse("~3.4")?.matches(&Version::parse(version)?) {
		Err(format!("OpenCV version: {} must be from 3.4 branch because of the feature: opencv-34", version).into())
	} else if cfg!(feature = "opencv-4") && !VersionReq::parse("~4")?.matches(&Version::parse(version)?) {
		Err(format!("OpenCV version: {} must be from 4.x branch because of the feature: opencv-4", version).into())
	} else {
		Ok(())
	}
}

fn get_versioned_hub_dirs() -> (PathBuf, PathBuf) {
	let bindings_dir = MANIFEST_DIR.join("bindings");
	let mut rust_hub_dir = bindings_dir.join("rust");
	let mut cpp_hub_dir = bindings_dir.join("cpp");
	if cfg!(feature = "opencv-32") {
		rust_hub_dir.push("opencv_32");
		cpp_hub_dir.push("opencv_32");
	} else if cfg!(feature = "opencv-34") {
		rust_hub_dir.push("opencv_34");
		cpp_hub_dir.push("opencv_34");
	} else if cfg!(feature = "opencv-4") {
		rust_hub_dir.push("opencv_4");
		cpp_hub_dir.push("opencv_4");
	}
	(rust_hub_dir, cpp_hub_dir)
}

fn make_modules(opencv_dir_as_string: &str) -> Result<()> {
	if MODULES.get().is_some() {
		return Ok(());
	}
	let ignore_modules: HashSet<&'static str> = HashSet::from_iter([
		"core_detect",
		"cudalegacy",
		"cudev",
		"gapi",
		"hal",
		"hfs",
		"ippicv",
		"opencv",
		"opencv_modules",
		"optflow",
		"quality",
		"rgbd",
		"saliency",
		"stereo",
		"tracking",
		"ximgproc",
	].iter().copied());

	let modules: Vec<String> = glob(&format!("{}/*.hpp", opencv_dir_as_string))?
		.filter_map(|entry| {
			let entry = entry.expect("Can't get path for module file");
			let module: String = entry.file_stem().expect("Can't calculate file stem")
				.to_string_lossy()
				.into_owned();
			if ignore_modules.contains(module.as_str()) {
				None
			} else {
				Some(module)
			}
		})
		.collect();

	MODULES.set(modules).expect("Can't set MODULES cache");
	Ok(())
}

fn build_compiler(opencv_header_dir: &Path) -> cc::Build {
	let mut out = cc::Build::new();
	out.cpp(true)
		.include(&*SRC_CPP_DIR)
		.include(opencv_header_dir)
		.include(&*OUT_DIR)
		.include(".")
		.flag_if_supported("-Wno-class-memaccess")
		.flag_if_supported("-Wno-deprecated-declarations")
		.flag_if_supported("-Wno-deprecated-copy")
		.flag_if_supported("-Wno-unused-variable")
		.flag_if_supported("-Wno-return-type-c-linkage")
	;
	if cfg!(target_env = "msvc") {
		out.flag_if_supported("-std:c++latest")
			.flag_if_supported("-wd4996")
			.flag_if_supported("-wd5054") // deprecated between enumerations of different types
			.flag_if_supported("-wd4190") // has C-linkage specified, but returns UDT 'Result<cv::Rect_<int>>' which is incompatible with C
			.flag_if_supported("-EHsc")
		;
	} else {
		out.flag("-std=c++11");
	}
	out
}

fn build_wrapper(opencv_header_dir: &Path) -> Result<()> {
	let modules = MODULES.get().expect("MODULES not initialized");

	let mut cc = build_compiler(opencv_header_dir);
	for module in modules {
		cc.file(OUT_DIR.join(format!("{}.cpp", module)));
		let manual_cpp = SRC_CPP_DIR.join(format!("manual-{}.cpp", module));
		if manual_cpp.exists() {
			cc.file(manual_cpp);
		}
	}
	cc.compile("ocvrs");
	Ok(())
}

fn install_wrapper() -> Result<()> {
	let (rust_hub_dir, cpp_hub_dir) = get_versioned_hub_dirs();
	let target_hub_dir = SRC_DIR.join("opencv");
	let target_module_dir = target_hub_dir.join("hub");

	for entry in glob(&format!("{}/*.cpp", cpp_hub_dir.to_str().unwrap()))? {
		let entry = entry?;
		file_copy_to_dir(&entry, &OUT_DIR)?;
	}
	for entry in glob(&format!("{}/*.h", cpp_hub_dir.to_str().unwrap()))? {
		let entry = entry?;
		file_copy_to_dir(&entry, &OUT_DIR)?;
	}
	for entry in glob(&format!("{}/*.hpp", cpp_hub_dir.to_str().unwrap()))? {
		let entry = entry?;
		file_copy_to_dir(&entry, &OUT_DIR)?;
	}

	for entry in glob(&format!("{}/*.rs", target_module_dir.to_str().unwrap()))? {
		let _ = fs::remove_file(entry?);
	}
	for entry in glob(&format!("{}/**/*.rs", rust_hub_dir.to_str().unwrap())).unwrap() {
		let entry = entry?;
		let target_file = target_hub_dir.join(entry.strip_prefix(&rust_hub_dir)?);
		if let Some(target_dir) = target_file.parent() {
			if !target_dir.exists() {
				fs::create_dir_all(target_dir)?;
			}
		}
		fs::copy(&entry, target_file)?;
	}
	Ok(())
}

fn cleanup() -> Result<()> {
	// fixme, shouldn't be needed
	for entry in glob(&format!("{}/*.rs", OUT_DIR.to_string_lossy()))? {
		let _ = fs::remove_file(entry?);
	}
	Ok(())
}

fn main() -> Result<()> {
	let features = [cfg!(feature = "opencv-32"), cfg!(feature = "opencv-34"), cfg!(feature = "opencv-4")].iter().map(|&x| i32::from(x)).sum::<i32>();
	if features != 1 {
		panic!("Please select exactly one of the features: opencv-32, opencv-34, opencv-4");
	}
	let opencv = if cfg!(feature = "docs-only") {
		Library::probe_from_paths("opencv", "", "", "")?
	} else {
		Library::probe()?
	};
	eprintln!("=== OpenCV library configuration: {:#?}", opencv);
	if !cfg!(feature = "docs-only") {
		check_matching_version(&opencv.version)
			.map_err(|e| format!("{}, (version coming from pkg_config for package: {})", e, opencv.pkg_name))?;
	}
	let opencv_header_dir = env::var_os("OPENCV_HEADER_DIR").map(PathBuf::from).unwrap_or_else(|| {
		if cfg!(feature = "buildtime-bindgen") {
			opencv.include_paths.iter()
				.find(|p| p.join("opencv2").exists())
				.expect("Using buildtime-bindgen, but discovered OpenCV include paths is empty or contains non-existent paths").clone()
		} else {
			if cfg!(feature = "opencv-32") {
				MANIFEST_DIR.join("headers/3.2")
			} else if cfg!(feature = "opencv-34") {
				MANIFEST_DIR.join("headers/3.4")
			} else if cfg!(feature = "opencv-4") {
				MANIFEST_DIR.join("headers/4")
			} else {
				panic!("Please select one OpenCV major version using one of the opencv-* features or specify OpenCV header path manually via OPENCV_HEADER_DIR environment var");
			}
		}
	});

	make_modules(&opencv_header_dir.join("opencv2").to_string_lossy())?;

	if let Some(version) = get_version_from_headers(&opencv_header_dir) {
		check_matching_version(&version).map_err(|e| format!("{}, (version coming from headers at: {})", e, opencv_header_dir.display()))?;
		eprintln!("=== Found OpenCV library version: {} in headers located at: {}", version, opencv_header_dir.display());
	} else {
		panic!("Unable to find header version in: {}", opencv_header_dir.display())
	}

	#[cfg(feature = "buildtime-bindgen")]
		generator::gen_wrapper(&opencv_header_dir)?;
	install_wrapper()?;
	build_wrapper(&opencv_header_dir)?;
	// -l linker args should be emitted after -l static
	opencv.emit_cargo_metadata();
	cleanup()?;
	Ok(())
}
